pipelines/trace_test.go

```diff
@@ -104,3 +109,54 @@ func TestSecureTrace(t *testing.T) {
 
 	require.Equal(t, []string{"test-value"}, server.TraceMDs()[0]["test-header"])
 }
+
+func TestSampledTrace(t *testing.T) {
+	server := test.NewServer()
+	defer server.Stop()
+
+	shutdown, err := NewTracePipeline(PipelineConfig{
+		Endpoint: fmt.Sprintf("%s:%d", test.ServerName, server.SecureTracePort),
+		Headers: map[string]string{
+			"test-header": "test-value",
+		},
+		Resource: resource.NewWithAttributes(
+			semconv.SchemaURL,
+			attribute.String("test-r1", "test-v1"),
+		),
+		ReportingPeriod: "24h",
+		Credentials:     credentials.NewTLS(newTLSConfig()),
+		Propagators:     []string{"tracecontext", "baggage"},
+		SamplingEnabled: true,
+		SamplingPercent: 50,
+	})
+	require.NoError(t, err)
+
+	tracer := otel.Tracer("test-library")
+	for i := 0; i < 100; i++ {
+		_, span := tracer.Start(context.Background(), ("test-span"))
+		span.End()
+	}
+
+	require.NoError(t, shutdown())
+
+	require.Equal(t, 0, len(server.MetricsRequests()))
+	require.Equal(t, 1, len(server.TraceRequests()))
+
+	nSpans := spanCount(server)
+	require.Greater(t, nSpans, 30)
```

Comment:
Whether a trace gets sampled or not is deterministic based on the trace ID, which is non-deterministic. So for a short run, like here, more or fewer than the expected 50 spans might be sampled.

I didn't see a great way to force the trace IDs to be deterministic here, so I opted for a fairly wide error range on the expectation instead. Technically this test is flaky, but I didn't get a single failure in 10,000 runs:

```
$ go test ./... -test.count=10000 -test.run=TestSampledTrace

ok  	github.com/lightstep/otel-launcher-go/pipelines	75.658s
?   	github.com/lightstep/otel-launcher-go/pipelines/test	[no test files]
```

If setting sampling fails, then exactly 100 spans will be created and this test will fail, of course.

Happy to consider alternative ways to test this.

Reply:
Consistent probability sampler means this can now be a specific number